torch/testing/_internal/common_methods_invocations.py

```diff
@@ -17375,6 +17425,34 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
         torch_opinfo_name="clamp",
         supports_nvfuser=False,
     ),
+    PythonRefInfo(
```

Comment:
So back to @lezcano's point about operators written in Python. Now that I've had a chance to review the PR let me elaborate on my previous comment and see if @lezcano agrees with this perspective.

The core idea of primTorch is that, ultimately, PyTorch programs can be represented using a smaller set of operators that more simply describe the program's semantics. So it's natural to write operations using those primitives and then operations using those operations and so on. 

PyTorch operations that are entirely defined in Python are effectively user-defined operations that live under the `torch` namespace. We shouldn't need to add explicit Python references for them because they are their own Python references. However, as we're seeing with many other PyTorch operators, there are often opportunities to improve these operations and/or their tests to be more consistent with the detailed attention we should be paying to Python references.

Because of that, I don't think we want to add these as Python references. Instead my question is: why not update the actual operator implementations with the changes made here? Does `_apply_loss_reduction` need to be updated to call `torch.sum` instead of `_refs.sum` directly? Are there other divergences? Is running the refs eagerly slower than the actual operation? Etc. As this PR points out there are fixes here for the original operations, and the test extension is nice, I just think we can drop addition of the `PythonRefInfo` and adding these operations to the `_refs` namespace. 

What are your thoughts, @nkaretnikov?

Reply:
As I see it, there's three options

1. [Mike's]: Not writing the ref at all, just have the OG Python version
2. [Extension of Mike's]: The ref should be an alias (`fn_alias = fn`) of the OG.
3. The main function should be in refs, and the function in the main namespace should be an alias to this function 

Between 1 and 2, I think that 2 is strictly better than 1. Also, everything that Mike mentioned applies to 2, so I'll just discuss 3.

So, the thing is that in `_refs` we do have some very nice tools developed to write operators in Python. `_apply_loss_reduction` and `utils.check` are two great examples of this.
As such, there may be some value in rewriting these operators as `_refs`. If we go down this route, then we surely will need to add matching `PythonRefInfo`, of course. If this is done an the ref includes improvements, these should also be reflected in the original version. To do this, we can simply write the original version as an alias to the ref (!!). This suggestion goes in line with the line of though of "PrimTorch will, at some point, become the main entry point to the PyTorch library".

An issue with this approach may be that we may run into circular imports, so I don't know how feasible is it.

Another issue related to the above is that we would start interweaving PrimTorch with core, which we may not want to do just yet.

The pros of "having all the implementations together in the same format" seems nice. For example, this function should probably need some broadcasting which it doesn't currently have. Now, the cons are also quite annoying, so I don't know. What do you guys think?

Reply:
Good idea that we may want an alias, @lezcano -- as you say, why not?

For now I would suggesting going with approach (2). It's easy to switch to approach (3) later if we want, as that's just about what file the function is defined in. Having a single implementation of these Python-only operations seems like the thing.

I wouldn't worry too much about "interweaving primTorch with core." The primTorch project effectively has distinct parts, including the primitives themselves and a utility library for writing PyTorch operations in Python. I wouldn't want these operations to call prims directly at this time, but because the references are written as calling into the torch namespace I think components of the utility library are reusable without issue (like the functions you mention).

Typically with operations like this that are defined as composites PyTorch hasn't explicitly defined their broadcasting or type promotion behavior, and relied on the behavior inherited from the composition. This is imperfect since our type promotion behavior isn't composable, but it's been good enough and I don't think we should seek to change that standard with this PR. Also, the type promotion decorator we use with the primTorch Python references is probably too slow to use here.

Reply:
> [Extension of Mike's]: The ref should be an alias (fn_alias = fn) of the OG.

I've tried this approach with `gaussian_nll_loss` from another PR that I have opened. It results in `maximum recursion depth exceeded` during testing (ref dispatches to orig impl which dispatches to ref ...). So I'd suggest to just skip adding references for pure Python ops unless we have a good reason to do it. We can always do it later since they all should be copy-pastable.

Personally, I'm not in favor of mixing regular PyTorch and primTorch just yet. We can do it later once the major issues are addressed. As of now, I feel like this mixing adds up to "maintenance cost" for no good reason.

So my vote is to go with this for now (while keeping a list of pure Python ops somewhere -- for later):
> [Mike's]: Not writing the ref at all, just have the OG Python version

Reply:
Just updating the original Python version with these improvements and the test extension sounds great for now